homeassistant/components/enasolar/sensor.py

```diff
@@ -0,0 +1,274 @@
+    CONF_HOST,
+    CONF_NAME,
+    DEVICE_CLASS_ENERGY,
+    DEVICE_CLASS_POWER,
+    DEVICE_CLASS_TEMPERATURE,
+    ENERGY_KILO_WATT_HOUR,
+    EVENT_HOMEASSISTANT_START,
+    POWER_KILO_WATT,
+    TEMP_CELSIUS,
+    TEMP_FAHRENHEIT,
+)
+from homeassistant.core import CALLBACK_TYPE, callback
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers.event import async_call_later
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    CONF_CAPABILITY,
+    CONF_DC_STRINGS,
+    CONF_MAX_OUTPUT,
+    CONF_SUN_DOWN,
+    CONF_SUN_UP,
+    DEFAULT_SUN_DOWN,
+    DEFAULT_SUN_UP,
+    DOMAIN,
+    ENASOLAR_UNIT_MAPPINGS,
+    SCAN_DATA_MIN_INTERVAL,
+    SCAN_MAX_INTERVAL,
+    SCAN_METERS_MIN_INTERVAL,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Add enasolar entry."""
+
+    # Use all sensors by default, but split them to have two update frequencies
+    hass_meter_sensors = []
+    hass_data_sensors = []
+    listeners = []
+
+    host = config_entry.data[CONF_HOST]
+
+    _LOGGER.info("Instantiate an EnaSolar Inverter at '%s'", host)
+    enasolar = pyenasolar.EnaSolar()
+
+    try:
+        await enasolar.interogate_inverter(host)
+    except Exception as conerr:
+        _LOGGER.error("Connection to EnaSolar Inverter '%s' failed (%s)", host, conerr)
+        raise PlatformNotReady from conerr
+
+    if config_entry.options != {}:
+        enasolar.sun_up = dt_util.parse_time(config_entry.options[CONF_SUN_UP])
+        enasolar.sun_down = dt_util.parse_time(config_entry.options[CONF_SUN_DOWN])
+    else:
+        enasolar.sun_up = dt_util.parse_time(DEFAULT_SUN_UP)
+        enasolar.sun_down = dt_util.parse_time(DEFAULT_SUN_DOWN)
+
+    enasolar.capability = config_entry.data[CONF_CAPABILITY]
+    enasolar.dc_strings = config_entry.data[CONF_DC_STRINGS]
+    enasolar.max_output = config_entry.data[CONF_MAX_OUTPUT]
+
+    _LOGGER.info("   Polling between %s and %s", enasolar.sun_up, enasolar.sun_down)
+    _LOGGER.info(
+        "   Max Output: %s, DC Strings: %s, Capability: %s",
+        enasolar.max_output,
+        enasolar.dc_strings,
+        enasolar.capability,
+    )
+
+    enasolar.setup_sensors()
+    for sensor in enasolar.sensors:
+        _LOGGER.debug("Setup sensor %s", sensor.key)
+        if sensor.enabled:
+            if sensor.is_meter:
+                hass_meter_sensors.append(
+                    EnaSolarSensor(
+                        sensor, config_entry.data[CONF_NAME], enasolar.serial_no
+                    )
+                )
+            else:
+                hass_data_sensors.append(
+                    EnaSolarSensor(
+                        sensor, config_entry.data[CONF_NAME], enasolar.serial_no
+                    )
+                )
+
+    async_add_entities(hass_meter_sensors)
+    async_add_entities(hass_data_sensors)
+
+    async def async_enasolar_meters():
```

Comment:
These look like these nearly reimplement the built-in `DataUpdateCoordinator` https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities

Reply:
This has been suggested already but I thought I would leave for a later version as the logic I currently have seems to work

Reply:
Deferring at this stage